### PR TITLE
Dispatchopts

### PR DESCRIPTION
**Pull request by RGunning**
Added option arguments to dispatch simple
set option in plite vsc

e.g.
dispatcher=dispatch-basic simple
dispatcher=dispatch-basic lsf -output this/dir

### CODE REVIEW

File: lib/App/Pipeline/Lite4.pm [hunk 1]
```diff
@@ -268,21 +268,20 @@ sub run_pipeline {
...
+    my $self = shift;
+    my $dispatcher_opt = shift;
+    my ($dispatcher_exe, $opts) = split / /, $dispatcher_opt;
     
     my $pipeline_graph_file = $self->pipeline_graph_file->absolute;
     
-    my $pipeline_dir = $self->pipeline_dir->absolute->stringify;
-    ouch 'App_Pipeline_Lite4_Error', "Dispatcher does not exist at $dispatcher_exe" 
-     unless $dispatcher_exe->exists;
+  # ouch 'App_Pipeline_Lite4_Error', "Dispatcher does not exist at $dispatcher_exe"
```
Comment (RGunning) [inline lib/App/Pipeline/Lite4.pm]: TODO: re-enable test if file exists

**Comment by nw11** [comment]
How much do you switch between dispatch-basic and lsf? I was thinking the best design for this, esp because more development might end up being done on the lsf dispatcher, is to have two dispatcher scripts dispatch-lsf and dispatch-basic. Then you just set them in the config file as you need (including setting the option to lsf for where it puts its log files). I don't think the external_dispatch function actually needs to be touched - it should just work fine. So basically, just paste the code for lsf stuff into dispatch-lsf.  I'm thinking this is the best strategy as I have some dispatcher scripts for job management systems besides lsf almost ready, and it seems more organised to do it this way, rather than to have a potentially massive file managing dispatch variants.

**Comment by RGunning** [comment]
I never really switch.
It would make more sense to have them all the dispatchers in separate files.

However, for lsf, it would be a useful feature to pass the output directory through an option passed via plite vsc. This is simpler than requiring the user to edit the dispatcher script to change output directory.
It was because I wanted to do this that I started editing the external_dispatch function

File: bin/dispatchers/dispatch-basic [hunk 2]
```diff
@@ -1,57 +1,157 @@
...
+   		my $mem = $graph_and_jobs->{ $job }->{$step}->{mem};
+   		my $queue = $graph_and_jobs->{ $job }->{$step}->{queue};
+   		my $cores = $graph_and_jobs->{ $job }->{$step}->{cores};
+   		my %args;
+   		$args{-e} = $errorfile if defined($errorfile);
+	 	$args{-w} = $dependency_str if @dependent_jobs >0;
+	 	$args{-J} = $lsf_job_name;
+		$args{-R} = "span[hosts=1] select[mem>$mem] rusage[mem=$mem]" if defined($mem);
+		$args{-M} = $mem if defined($mem); # in KB, but we * by 1000 not 1024, since esub requires it to be less.
+		$args{-o} = $c->options->{'output'}.'/out-%J';
```
Comment (RGunning) [inline bin/dispatchers/dispatch-basic]: set output directory based on options.
Should probably set a default value

**Comment by nw11** [comment]
Yes it seems we are essentially talking about the same thing here.   

With something like this in the system config file (which is brought up using plite vsc): 

dispatcher=dispatch-lsf  -output this/dir

This would just be executed fine as per normal by external_dispatch. As it is just going to join the command ("dispatch-lsf -output this/dir") to the actual lsf commands file to get "dispatch-lsf -output this/dir some-lsf-cmd-graph-file.yaml", which is then run by system() in external_dispatch.

**Comment by RGunning** [comment]
Not quite.

If you notice the line I commented out in the external_dispatch, the function is trying to test if the file at "dispatch-lsf -output this/dir” exists. This fails as only the first half of the command is a filepath, the second are options. My edits just separate out the file so I can test whether the file exists/executable

File: lib/App/Pipeline/Lite4.pm [hunk 3]
```diff
@@ -268,21 +268,20 @@ sub run_pipeline {
 
 sub external_dispatch {
    # TYPE  Path::Tiny :$dispatcher_exe )  {
-    my $self=shift;
-    my $dispatcher_exe=shift;
+    my $self = shift;
+    my $dispatcher_opt = shift;
+    my ($dispatcher_exe, $opts) = split / /, $dispatcher_opt;
```
Comment (nw11) [inline lib/App/Pipeline/Lite4.pm]: probably do /\s+/ to make a bit more robust

**Comment by nw11** [comment]
ah yes, of course. Should have looked more closely! ok good, this should be fine  to merge, when the two dispatchers are separated

**Comment by RGunning** [comment]
I've split the dispatchers.

I should probably look at uncommenting lines l277-279 of external_dispatch which tests if the dispatcher file exists. Although l282 also does this when it tests if the dispatcher is executable

**Pull request closed without merging**